// @aps_sdk/oss/dist/api/objectsApi.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ObjectsApi = exports.ObjectsApiFp = exports.ObjectsApiAxiosParamCreator = void 0;
const autodesk_sdkmanager_1 = require("@aps_sdk/autodesk-sdkmanager");
const common_1 = require("../common");
const base_1 = require("../base");
/**
 * ObjectsApi - axios parameter creator
 * @export
 */
const ObjectsApiAxiosParamCreator = function (apsConfiguration) {
    return {
        /**
         * Requests OSS to start reconstituting the set of objects that were uploaded using signed S3 upload URLs. You must call this operation only after all the objects have been uploaded.   You can specify up to 25 objects in this operation.
         * @summary Complete Batch Upload to S3 Signed URLs
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {BatchcompleteuploadObject} [requests] An array of objects, each of which represents an upload to complete.
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        batchCompleteUpload: (accessToken_1, bucketKey_1, requests_1, ...args_1) => __awaiter(this, [accessToken_1, bucketKey_1, requests_1, ...args_1], void 0, function* (accessToken, bucketKey, requests, options = {}) {
            // verify required parameter 'bucketKey' is not null or undefined
            (0, common_1.assertParamExists)('batchCompleteUpload', 'bucketKey', bucketKey);
            const localVarPath = `/oss/v2/buckets/{bucketKey}/objects/batchcompleteupload`
                .replace(`{${"bucketKey"}}`, encodeURIComponent(String(bucketKey)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            localVarHeaderParameter['Content-Type'] = 'application/json';
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(requests, localVarRequestOptions, apsConfiguration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Creates and returns signed URLs to download a set of objects directly from S3. These signed URLs expire in 2 minutes by default, but you can change this duration if needed.  You must start download the objects before the signed URLs expire. The download itself can take longer.  Only the application that owns the bucket can call this operation. All other applications that call this operation will receive a \"403 Forbidden\" error.
         * @summary Batch Generate Signed S3 Download URLs
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {Batchsigneds3downloadObject} requests An array of objects representing each request for a signed download URL.
         * @param {boolean} [publicResourceFallback] Specifies how to return the signed URLs, in case the object was uploaded in chunks, and assembling of chunks is not yet complete.  - &#x60;&#x60;true&#x60;&#x60; : Return a single signed OSS URL. - &#x60;&#x60;false&#x60;&#x60; : (Default) Return multiple signed S3 URLs, where each URL points to a chunk.
         * @param {number} [minutesExpiration] The time window (in minutes) the signed URL will remain usable. Acceptable values &#x3D; 1-60 minutes. Default &#x3D; 2 minutes.  **Tip:** Use the smallest possible time window to minimize exposure of the signed URL.
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        batchSignedS3Download: (accessToken_1, bucketKey_1, requests_1, publicResourceFallback_1, minutesExpiration_1, ...args_1) => __awaiter(this, [accessToken_1, bucketKey_1, requests_1, publicResourceFallback_1, minutesExpiration_1, ...args_1], void 0, function* (accessToken, bucketKey, requests, publicResourceFallback, minutesExpiration, options = {}) {
            // verify required parameter 'bucketKey' is not null or undefined
            (0, common_1.assertParamExists)('batchSignedS3Download', 'bucketKey', bucketKey);
            // verify required parameter 'requests' is not null or undefined
            (0, common_1.assertParamExists)('batchSignedS3Download', 'requests', requests);
            const localVarPath = `/oss/v2/buckets/{bucketKey}/objects/batchsigneds3download`
                .replace(`{${"bucketKey"}}`, encodeURIComponent(String(bucketKey)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            if (publicResourceFallback !== undefined) {
                localVarQueryParameter['public-resource-fallback'] = publicResourceFallback;
            }
            if (minutesExpiration !== undefined) {
                localVarQueryParameter['minutesExpiration'] = minutesExpiration;
            }
            localVarHeaderParameter['Content-Type'] = 'application/json';
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(requests, localVarRequestOptions, apsConfiguration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Creates and returns signed URLs to upload a set of objects directly to S3. These signed URLs expire in 2 minutes by default, but you can change this duration if needed.  You must start uploading the objects before the signed URLs expire. The upload  itself can take longer.  Only the application that owns the bucket can call this operation. All other applications that call this operation will receive a \"403 Forbidden\" error.  If required, you can request an array of signed URLs for each object, which lets you upload the objects in chunks. Once you upload all chunks you must call the [Complete Batch Upload to S3 Signed URL](/en/docs/data/v2/reference/http/buckets-:bucketKey-objects-:objectKey-batchcompleteupload-POST/) operation to indicate completion. This instructs OSS to assemble the chunks and reconstitute the object on OSS. You must call this operation even if you requested a single signed URL for an object.  If an upload fails after the validity period of a signed URL has elapsed, you can call this operation again to obtain fresh signed URLs. However, you must use the same ``uploadKey`` that was returned when you originally called this operation.
         * @summary Batch Generate Signed S3 Upload URLs
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {boolean} [useAcceleration] &#x60;&#x60;true&#x60;&#x60; : (Default) Generates a faster S3 signed URL using Transfer Acceleration.  &#x60;&#x60;false&#x60;&#x60;: Generates a standard S3 signed URL.
         * @param {number} [minutesExpiration] The time window (in minutes) the signed URL will remain usable. Acceptable values &#x3D; 1-60 minutes. Default &#x3D; 2 minutes.  **Tip:** Use the smallest possible time window to minimize exposure of the signed URL.
         * @param {Batchsigneds3uploadObject} [requests] An array of objects representing each request for a signed upload URL.
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        batchSignedS3Upload: (accessToken_1, bucketKey_1, useAcceleration_1, minutesExpiration_1, requests_1, ...args_1) => __awaiter(this, [accessToken_1, bucketKey_1, useAcceleration_1, minutesExpiration_1, requests_1, ...args_1], void 0, function* (accessToken, bucketKey, useAcceleration, minutesExpiration, requests, options = {}) {
            // verify required parameter 'bucketKey' is not null or undefined
            (0, common_1.assertParamExists)('batchSignedS3Upload', 'bucketKey', bucketKey);
            const localVarPath = `/oss/v2/buckets/{bucketKey}/objects/batchsigneds3upload`
                .replace(`{${"bucketKey"}}`, encodeURIComponent(String(bucketKey)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            if (useAcceleration !== undefined) {
                localVarQueryParameter['useAcceleration'] = useAcceleration;
            }
            if (minutesExpiration !== undefined) {
                localVarQueryParameter['minutesExpiration'] = minutesExpiration;
            }
            localVarHeaderParameter['Content-Type'] = 'application/json';
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(requests, localVarRequestOptions, apsConfiguration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Requests OSS to assemble and reconstitute the object that was uploaded using signed S3 upload URL. You must call this operation only after all parts/chunks of the object has been uploaded.
         * @summary Complete Upload to S3 Signed URL
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {string} contentType Must be &#x60;&#x60;application/json&#x60;&#x60;.
         * @param {Completes3uploadBody} body
         * @param {string} [xAdsMetaContentType] The Content-Type value for the uploaded object to record within OSS.
         * @param {string} [xAdsMetaContentDisposition] The Content-Disposition value for the uploaded object to record within OSS.
         * @param {string} [xAdsMetaContentEncoding] The Content-Encoding value for the uploaded object to record within OSS.
         * @param {string} [xAdsMetaCacheControl] The Cache-Control value for the uploaded object to record within OSS.
         * @param {string} [xAdsUserDefinedMetadata] Custom metadata to be stored with the object, which can be retrieved later on download or when retrieving object details. Must be a JSON object that is less than 100 bytes.
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        completeSignedS3Upload: (accessToken_1, bucketKey_1, objectKey_1, contentType_1, body_1, xAdsMetaContentType_1, xAdsMetaContentDisposition_1, xAdsMetaContentEncoding_1, xAdsMetaCacheControl_1, xAdsUserDefinedMetadata_1, ...args_1) => __awaiter(this, [accessToken_1, bucketKey_1, objectKey_1, contentType_1, body_1, xAdsMetaContentType_1, xAdsMetaContentDisposition_1, xAdsMetaContentEncoding_1, xAdsMetaCacheControl_1, xAdsUserDefinedMetadata_1, ...args_1], void 0, function* (accessToken, bucketKey, objectKey, contentType, body, xAdsMetaContentType, xAdsMetaContentDisposition, xAdsMetaContentEncoding, xAdsMetaCacheControl, xAdsUserDefinedMetadata, options = {}) {
            // verify required parameter 'bucketKey' is not null or undefined
            (0, common_1.assertParamExists)('completeSignedS3Upload', 'bucketKey', bucketKey);
            // verify required parameter 'objectKey' is not null or undefined
            (0, common_1.assertParamExists)('completeSignedS3Upload', 'objectKey', objectKey);
            // verify required parameter 'contentType' is not null or undefined
            (0, common_1.assertParamExists)('completeSignedS3Upload', 'contentType', contentType);
            // verify required parameter 'body' is not null or undefined
            (0, common_1.assertParamExists)('completeSignedS3Upload', 'body', body);
            const localVarPath = `/oss/v2/buckets/{bucketKey}/objects/{objectKey}/signeds3upload`
                .replace(`{${"bucketKey"}}`, encodeURIComponent(String(bucketKey)))
                .replace(`{${"objectKey"}}`, encodeURIComponent(String(objectKey)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            if (contentType != null) {
                localVarHeaderParameter['Content-Type'] = String(contentType);
            }
            if (xAdsMetaContentType != null) {
                localVarHeaderParameter['x-ads-meta-Content-Type'] = String(xAdsMetaContentType);
            }
            if (xAdsMetaContentDisposition != null) {
                localVarHeaderParameter['x-ads-meta-Content-Disposition'] = String(xAdsMetaContentDisposition);
            }
            if (xAdsMetaContentEncoding != null) {
                localVarHeaderParameter['x-ads-meta-Content-Encoding'] = String(xAdsMetaContentEncoding);
            }
            if (xAdsMetaCacheControl != null) {
                localVarHeaderParameter['x-ads-meta-Cache-Control'] = String(xAdsMetaCacheControl);
            }
            if (xAdsUserDefinedMetadata != null) {
                localVarHeaderParameter['x-ads-user-defined-metadata'] = String(xAdsUserDefinedMetadata);
            }
            localVarHeaderParameter['Content-Type'] = 'application/json';
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(body, localVarRequestOptions, apsConfiguration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Creates a copy of an object within the bucket.
         * @summary Copy Object
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {string} newObjName A URL-encoded human friendly name to identify the copied object.
         * @param {string} [xAdsAcmNamespace] This header is used to let the OSS Api Proxy know if ACM is used to authorize access to the given object. If this authorization is used by your service, then you must provide the name of the namespace you want to validate access control policies against.
         * @param {string} [xAdsAcmCheckGroups] Informs the OSS Api Proxy know if your service requires ACM authorization to also validate against Oxygen groups. If so, you must pass this header with a value of &#x60;&#x60;true&#x60;&#x60;. Otherwise, the assumption is that checking authorization against Oxygen groups is not required.
         * @param {string} [xAdsAcmGroups] Use this header to pass the Oxygen groups you want the OSS Api Proxy to use for group validation for the given user in the OAuth2 token.
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        copyTo: (accessToken_1, bucketKey_1, objectKey_1, newObjName_1, ...args_1) => __awaiter(this, [accessToken_1, bucketKey_1, objectKey_1, newObjName_1, ...args_1], void 0, function* (accessToken, bucketKey, objectKey, newObjName, options = {}) {
            // verify required parameter 'bucketKey' is not null or undefined
            (0, common_1.assertParamExists)('copyTo', 'bucketKey', bucketKey);
            // verify required parameter 'objectKey' is not null or undefined
            (0, common_1.assertParamExists)('copyTo', 'objectKey', objectKey);
            // verify required parameter 'newObjName' is not null or undefined
            (0, common_1.assertParamExists)('copyTo', 'newObjName', newObjName);
            const localVarPath = `/oss/v2/buckets/{bucketKey}/objects/{objectKey}/copyto/{newObjName}`
                .replace(`{${"bucketKey"}}`, encodeURIComponent(String(bucketKey)))
                .replace(`{${"objectKey"}}`, encodeURIComponent(String(objectKey)))
                .replace(`{${"newObjName"}}`, encodeURIComponent(String(newObjName)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'PUT' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Generates a signed URL that can be used to download or upload an object within the specified expiration time. If the object the signed URL points to is deleted or expires before the signed URL expires, the signed URL will no longer be valid.  In addition to this operation that generates OSS signed URLs, OSS provides operations to generate S3 signed URLs. S3 signed URLs allow direct upload/download from S3 but are restricted to bucket owners. OSS signed URLs also allow upload/download and can be configured for access by other applications, making them suitable for sharing objects across applications.  Only the application that owns the bucket containing the object can call this operation.
         * @summary Generate OSS Signed URL
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {Access} [access]
         * @param {boolean} [useCdn] &#x60;&#x60;true&#x60;&#x60; : Returns a Cloudfront URL to the object instead of an Autodesk URL (one that points to a location on https://developer.api.autodesk.com). Applications can interact with the Cloudfront URL exactly like with any classic public resource in OSS. They can use GET requests to download objects or PUT requests to upload objects.  &#x60;&#x60;false&#x60;&#x60; : (Default) Returns an Autodesk URL (one that points to a location on https://developer.api.autodesk.com) to the object.
         * @param {CreateSignedResource} [createSignedResource]
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createSignedResource: (accessToken_1, bucketKey_1, objectKey_1, access_1, useCdn_1, createSignedResource_1, ...args_1) => __awaiter(this, [accessToken_1, bucketKey_1, objectKey_1, access_1, useCdn_1, createSignedResource_1, ...args_1], void 0, function* (accessToken, bucketKey, objectKey, access, useCdn, createSignedResource, options = {}) {
            // verify required parameter 'bucketKey' is not null or undefined
            (0, common_1.assertParamExists)('createSignedResource', 'bucketKey', bucketKey);
            // verify required parameter 'objectKey' is not null or undefined
            (0, common_1.assertParamExists)('createSignedResource', 'objectKey', objectKey);
            const localVarPath = `/oss/v2/buckets/{bucketKey}/objects/{objectKey}/signed`
                .replace(`{${"bucketKey"}}`, encodeURIComponent(String(bucketKey)))
                .replace(`{${"objectKey"}}`, encodeURIComponent(String(objectKey)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'POST' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            if (access !== undefined) {
                localVarQueryParameter['access'] = access;
            }
            if (useCdn !== undefined) {
                localVarQueryParameter['useCdn'] = useCdn;
            }
            localVarHeaderParameter['Content-Type'] = 'application/json';
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = (0, common_1.serializeDataIfNeeded)(createSignedResource, localVarRequestOptions, apsConfiguration);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Deletes an object from the bucket.
         * @summary Delete Object
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {string} [xAdsAcmNamespace] This header is used to let the OSS Api Proxy know if ACM is used to authorize access to the given object. If this authorization is used by your service, then you must provide the name of the namespace you want to validate access control policies against.
         * @param {string} [xAdsAcmCheckGroups] Informs the OSS API Proxy know if your service requires ACM authorization to also validate against Oxygen groups. If so, you must pass this header with a value of &#x60;&#x60;true&#x60;&#x60;. Otherwise, the assumption is that checking authorization against Oxygen groups is not required.
         * @param {string} [xAdsAcmGroups] Use this header to pass the Oxygen groups you want the OSS Api Proxy to use for group validation for the given user in the OAuth2 token.
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteObject: (accessToken_1, bucketKey_1, objectKey_1, ...args_1) => __awaiter(this, [accessToken_1, bucketKey_1, objectKey_1, ...args_1], void 0, function* (accessToken, bucketKey, objectKey, options = {}) {
            // verify required parameter 'bucketKey' is not null or undefined
            (0, common_1.assertParamExists)('deleteObject', 'bucketKey', bucketKey);
            // verify required parameter 'objectKey' is not null or undefined
            (0, common_1.assertParamExists)('deleteObject', 'objectKey', objectKey);
            const localVarPath = `/oss/v2/buckets/{bucketKey}/objects/{objectKey}`
                .replace(`{${"bucketKey"}}`, encodeURIComponent(String(bucketKey)))
                .replace(`{${"objectKey"}}`, encodeURIComponent(String(objectKey)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'DELETE' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Delete an object using an OSS signed URL to access it.  Only applications that own the bucket containing the object can call this operation.
         * @summary Delete Object Using Signed URL
         * @param {string} hash The ID component of the signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains &#x60;&#x60;hash&#x60;&#x60; as a URI parameter.
         * @param {Region} [xAdsRegion] Specifies where the bucket containing the object stored. Possible values are:  - &#x60;&#x60;US&#x60;&#x60; - (Default) Data center for the US region. - &#x60;&#x60;EMEA&#x60;&#x60; - Data center for the European Union, Middle East, and Africa. - &#x60;&#x60;APAC&#x60;&#x60; -  (Beta) Data center for Australia.  **Note:** Beta features are subject to change. Please do not use in production environments.
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteSignedResource: (accessToken_1, hash_1, xAdsRegion_1, ...args_1) => __awaiter(this, [accessToken_1, hash_1, xAdsRegion_1, ...args_1], void 0, function* (accessToken, hash, xAdsRegion, options = {}) {
            // verify required parameter 'hash' is not null or undefined
            (0, common_1.assertParamExists)('deleteSignedResource', 'hash', hash);
            const localVarPath = `/oss/v2/signedresources/{hash}`
                .replace(`{${"hash"}}`, encodeURIComponent(String(hash)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'DELETE' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            if (xAdsRegion != null) {
                localVarHeaderParameter['x-ads-region'] = typeof xAdsRegion === 'string'
                    ? xAdsRegion
                    : JSON.stringify(xAdsRegion);
            }
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Returns detailed information about the specified object.
         * @summary Get Object Details
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {string} [ifModifiedSince] A timestamp in the HTTP date format (Mon, DD Month YYYY HH:MM:SS GMT). The requested data is returned only if the object has been modified since the specified timestamp. If not, a 304 (Not Modified) HTTP status is returned.
         * @param {string} [xAdsAcmNamespace] This header is used to let the OSS Api Proxy know if ACM is used to authorize access to the given object. If this authorization is used by your service, then you must provide the name of the namespace you want to validate access control policies against.
         * @param {string} [xAdsAcmCheckGroups] Informs the OSS Api Proxy know if your service requires ACM authorization to also validate against Oxygen groups. If so, you must pass this header with a value of &#x60;&#x60;true&#x60;&#x60;. Otherwise, the assumption is that checking authorization against Oxygen groups is not required.
         * @param {string} [xAdsAcmGroups] Use this header to pass the Oxygen groups you want the OSS Api Proxy to use for group validation for the given user in the OAuth2 token.
         * @param {With} [_with]
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getObjectDetails: (accessToken_1, bucketKey_1, objectKey_1, ifModifiedSince_1, _with_1, ...args_1) => __awaiter(this, [accessToken_1, bucketKey_1, objectKey_1, ifModifiedSince_1, _with_1, ...args_1], void 0, function* (accessToken, bucketKey, objectKey, ifModifiedSince, _with, options = {}) {
            // verify required parameter 'bucketKey' is not null or undefined
            (0, common_1.assertParamExists)('getObjectDetails', 'bucketKey', bucketKey);
            // verify required parameter 'objectKey' is not null or undefined
            (0, common_1.assertParamExists)('getObjectDetails', 'objectKey', objectKey);
            const localVarPath = `/oss/v2/buckets/{bucketKey}/objects/{objectKey}/details`
                .replace(`{${"bucketKey"}}`, encodeURIComponent(String(bucketKey)))
                .replace(`{${"objectKey"}}`, encodeURIComponent(String(objectKey)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            if (_with !== undefined) {
                localVarQueryParameter['with'] = _with;
            }
            if (ifModifiedSince != null) {
                localVarHeaderParameter['If-Modified-Since'] = typeof ifModifiedSince === 'string'
                    ? ifModifiedSince
                    : JSON.stringify(ifModifiedSince);
            }
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Returns a list objects in the specified bucket.   Only the application that owns the bucket can call this operation. All other applications that call this operation will receive a \"403 Forbidden\" error.
         * @summary List Objects
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {number} [limit] The number of items you want per page. Acceptable values &#x3D; 1-100. Default &#x3D; 10.
         * @param {string} [beginsWith] Filters the results by the value you specify. Only the objects with their &#x60;&#x60;objectKey&#x60;&#x60; beginning with the specified string are returned.
         * @param {string} [startAt] The ID of the last item that was returned in the previous result set.  It enables the system to return subsequent items starting from the next one after the specified ID.
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getObjects: (accessToken_1, bucketKey_1, limit_1, beginsWith_1, startAt_1, ...args_1) => __awaiter(this, [accessToken_1, bucketKey_1, limit_1, beginsWith_1, startAt_1, ...args_1], void 0, function* (accessToken, bucketKey, limit, beginsWith, startAt, options = {}) {
            // verify required parameter 'bucketKey' is not null or undefined
            (0, common_1.assertParamExists)('getObjects', 'bucketKey', bucketKey);
            const localVarPath = `/oss/v2/buckets/{bucketKey}/objects`
                .replace(`{${"bucketKey"}}`, encodeURIComponent(String(bucketKey)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            if (limit !== undefined) {
                localVarQueryParameter['limit'] = limit;
            }
            if (beginsWith !== undefined) {
                localVarQueryParameter['beginsWith'] = beginsWith;
            }
            if (startAt !== undefined) {
                localVarQueryParameter['startAt'] = startAt;
            }
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Downloads an object using an OSS signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/)  contains the ``hash`` URI parameter as well.
         * @summary Download Object Using Signed URL
         * @param {string} hash The ID component of the signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains &#x60;&#x60;hash&#x60;&#x60; as a URI parameter.
         * @param {string} [range] The byte range to download, specified in the form &#x60;&#x60;bytes&#x3D;&lt;START_BYTE&gt;-&lt;END_BYTE&gt;&#x60;&#x60;.
         * @param {string} [ifNoneMatch] The last known ETag value of the object. OSS returns the requested data  only if the &#x60;&#x60;If-None-Match&#x60;&#x60; header differs from the ETag value of the object on OSS, which indicates that the object on OSS is newer. If not, it returns a 304 \&quot;Not Modified\&quot; HTTP status.
         * @param {string} [ifModifiedSince] A timestamp in the HTTP date format (Mon, DD Month YYYY HH:MM:SS GMT). The requested data is returned only if the object has been modified since the specified timestamp. If not, a 304 (Not Modified) HTTP status is returned.
         * @param {string} [acceptEncoding] The compression format your prefer to receive the data. Possible values are:  - &#x60;&#x60;gzip&#x60;&#x60; - Use the gzip format  **Note:** You cannot use &#x60;&#x60;Accept-Encoding:gzip&#x60;&#x60; with a Range header containing an end byte range. OSS will not honor the End byte range if &#x60;&#x60;Accept-Encoding: gzip&#x60;&#x60; header is used.
         * @param {Region} [region] Specifies where the bucket containing the object stored. Possible values are:  - &#x60;&#x60;US&#x60;&#x60; - (Default) Data center for the US region. - &#x60;&#x60;EMEA&#x60;&#x60; - Data center for the European Union, Middle East, and Africa. - &#x60;&#x60;APAC&#x60;&#x60; -  (Beta) Data center for Australia.  **Note:** Beta features are subject to change. Please do not use in production environments.
         * @param {string} [responseContentDisposition] The value of the Content-Disposition header you want to receive when you download the object using the signed URL. If you do not specify a value, the Content-Disposition header defaults to the value stored with OSS.
         * @param {string} [responseContentType] The value of the Content-Type header you want to receive when you download the object using the signed URL. If you do not specify a value, the Content-Type header defaults to the value stored with OSS.
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSignedResource: (accessToken_1, hash_1, range_1, ifNoneMatch_1, ifModifiedSince_1, acceptEncoding_1, region_1, responseContentDisposition_1, responseContentType_1, ...args_1) => __awaiter(this, [accessToken_1, hash_1, range_1, ifNoneMatch_1, ifModifiedSince_1, acceptEncoding_1, region_1, responseContentDisposition_1, responseContentType_1, ...args_1], void 0, function* (accessToken, hash, range, ifNoneMatch, ifModifiedSince, acceptEncoding, region, responseContentDisposition, responseContentType, options = {}) {
            // verify required parameter 'hash' is not null or undefined
            (0, common_1.assertParamExists)('getSignedResource', 'hash', hash);
            const localVarPath = `/oss/v2/signedresources/{hash}`
                .replace(`{${"hash"}}`, encodeURIComponent(String(hash)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            if (region !== undefined) {
                localVarQueryParameter['region'] = region;
            }
            if (responseContentDisposition !== undefined) {
                localVarQueryParameter['response-content-disposition'] = responseContentDisposition;
            }
            if (responseContentType !== undefined) {
                localVarQueryParameter['response-content-type'] = responseContentType;
            }
            if (range != null) {
                localVarHeaderParameter['Range'] = String(range);
            }
            if (ifNoneMatch != null) {
                localVarHeaderParameter['If-None-Match'] = String(ifNoneMatch);
            }
            if (ifModifiedSince != null) {
                localVarHeaderParameter['If-Modified-Since'] = typeof ifModifiedSince === 'string'
                    ? ifModifiedSince
                    : JSON.stringify(ifModifiedSince);
            }
            if (acceptEncoding != null) {
                localVarHeaderParameter['Accept-Encoding'] = String(acceptEncoding);
            }
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Gets a signed URL to download an object directly from S3, bypassing OSS servers. This signed URL expires in 2 minutes by default, but you can change this duration if needed.  You must start the download before the signed URL expires. The download itself can take longer. If the download fails after the validity period of the signed URL has elapsed, you can call this operation again to obtain a fresh signed URL.  Only applications that have read access to the object can call this operation.     You can use range headers with the signed download URL to download the object in chunks. This ability lets you download chunks in parallel, which can result in faster downloads.  If the object you want to download was uploaded in chunks and is still assembling on OSS, you will receive multiple S3 URLs instead of just one. Each URL will point to a chunk. If you prefer to receive a single URL, set the ``public-resource-fallback`` query parameter to ``true``. This setting will make OSS fallback to returning a single signed OSS URL, if assembling is still in progress.   In addition to this operation that generates S3 signed URLs, OSS provides an operation to generate OSS signed URLs. S3 signed URLs allow direct upload/download from S3 but are restricted to bucket owners. OSS signed URLs also allow upload/download and can be configured for access by other applications, making them suitable for sharing objects across applications.
         * @summary Generate Signed S3 Download URL
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {string} [ifNoneMatch] The last known ETag value of the object. OSS returns the signed URL only if the &#x60;&#x60;If-None-Match&#x60;&#x60; header differs from the ETag value of the object on S3. If not, it returns a 304 \&quot;Not Modified\&quot; HTTP status.
         * @param {string} [ifModifiedSince] A timestamp in the HTTP date format (Mon, DD Month YYYY HH:MM:SS GMT). The signed URL is returned only if the object has been modified since the specified timestamp. If not, a 304 (Not Modified) HTTP status is returned.
         * @param {string} [xAdsAcmScopes] Optional OSS-compliant scope reference to increase bucket search performance
         * @param {string} [responseContentType] The value of the Content-Type header you want to receive when you download the object using the signed URL. If you do not specify a value, the Content-Type header defaults to the value stored with OSS.
         * @param {string} [responseContentDisposition] The value of the Content-Disposition header you want to receive when you download the object using the signed URL. If you do not specify a value, the Content-Disposition header defaults to the value stored with OSS.
         * @param {string} [responseCacheControl] The value of the Cache-Control header you want to receive when you download the object using the signed URL. If you do not specify a value, the Cache-Control header defaults to the value stored with OSS.
         * @param {boolean} [publicResourceFallback] Specifies how to return the signed URLs, in case the object was uploaded in chunks, and assembling of chunks is not yet complete.  - &#x60;&#x60;true&#x60;&#x60; : Return a single signed OSS URL. - &#x60;&#x60;false&#x60;&#x60; : (Default) Return multiple signed S3 URLs, where each URL points to a chunk.
         * @param {number} [minutesExpiration] The time window (in minutes) the signed URL will remain usable. Acceptable values &#x3D; 1-60 minutes. Default &#x3D; 2 minutes.  **Tip:** Use the smallest possible time window to minimize exposure of the signed URL.
         * @param {boolean} [useCdn] &#x60;&#x60;true&#x60;&#x60; : Returns a URL that points to a CloudFront edge location.  &#x60;&#x60;false&#x60;&#x60; : (Default) Returns a URL that points directly to the S3 object.
         * @param {boolean} [redirect] Generates a HTTP redirection response (Temporary Redirect 307​) using the generated URL.
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        signedS3Download: (accessToken_1, bucketKey_1, objectKey_1, ifNoneMatch_1, ifModifiedSince_1, responseContentType_1, responseContentDisposition_1, responseCacheControl_1, publicResourceFallback_1, minutesExpiration_1, useCdn_1, ...args_1) => __awaiter(this, [accessToken_1, bucketKey_1, objectKey_1, ifNoneMatch_1, ifModifiedSince_1, responseContentType_1, responseContentDisposition_1, responseCacheControl_1, publicResourceFallback_1, minutesExpiration_1, useCdn_1, ...args_1], void 0, function* (accessToken, bucketKey, objectKey, ifNoneMatch, ifModifiedSince, responseContentType, responseContentDisposition, responseCacheControl, publicResourceFallback, minutesExpiration, useCdn, options = {}) {
            // verify required parameter 'bucketKey' is not null or undefined
            (0, common_1.assertParamExists)('signedS3Download', 'bucketKey', bucketKey);
            // verify required parameter 'objectKey' is not null or undefined
            (0, common_1.assertParamExists)('signedS3Download', 'objectKey', objectKey);
            const localVarPath = `/oss/v2/buckets/{bucketKey}/objects/{objectKey}/signeds3download`
                .replace(`{${"bucketKey"}}`, encodeURIComponent(String(bucketKey)))
                .replace(`{${"objectKey"}}`, encodeURIComponent(String(objectKey)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            if (responseContentType !== undefined) {
                localVarQueryParameter['response-content-type'] = responseContentType;
            }
            if (responseContentDisposition !== undefined) {
                localVarQueryParameter['response-content-disposition'] = responseContentDisposition;
            }
            if (responseCacheControl !== undefined) {
                localVarQueryParameter['response-cache-control'] = responseCacheControl;
            }
            if (publicResourceFallback !== undefined) {
                localVarQueryParameter['public-resource-fallback'] = publicResourceFallback;
            }
            if (minutesExpiration !== undefined) {
                localVarQueryParameter['minutesExpiration'] = minutesExpiration;
            }
            if (useCdn !== undefined) {
                localVarQueryParameter['useCdn'] = useCdn;
            }
            if (ifNoneMatch != null) {
                localVarHeaderParameter['If-None-Match'] = String(ifNoneMatch);
            }
            if (ifModifiedSince != null) {
                localVarHeaderParameter['If-Modified-Since'] = typeof ifModifiedSince === 'string'
                    ? ifModifiedSince
                    : JSON.stringify(ifModifiedSince);
            }
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Gets a signed URL to upload an object directly to S3, bypassing OSS servers. You can also request an array of signed URLs which lets you upload an object in chunks.  This signed URL expires in 2 minutes by default, but you can change this duration if needed.  You must start the upload before the signed URL expires. The upload itself can take longer. If the upload fails after the validity period of the signed URL has elapsed, you can call this operation again to obtain a fresh signed URL (or an array of signed URLs as the case may be). However, you must use the same ``uploadKey`` that was returned when you originally called this operation.   Only applications that own the bucket can call this operation.  **Note:** Once you upload all chunks you must call the [Complete Upload to S3 Signed URL](/en/docs/data/v2/reference/http/buckets-:bucketKey-objects-:objectKey-signeds3upload-POST/) operation to indicate completion. This instructs OSS to assemble the chunks and reconstitute the object on OSS. You must call this operation even when using a single signed URL.   In addition to this operation that generates S3 signed URLs, OSS provides an operation to generate OSS signed URLs. S3 signed URLs allow direct upload/download from S3 but are restricted to bucket owners. OSS signed URLs also allow upload/download and can be configured for access by other applications, making them suitable for sharing objects across applications.
         * @summary Generate Signed S3 Upload URL
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {string} [xAdsAcmScopes] Optional OSS-compliant scope reference to increase bucket search performance
         * @param {number} [parts] The number of parts you intend to chunk the object for uploading. OSS will return that many signed URLs, one URL for each chunk. If you do not specify a value you\&#39;ll get only one URL to upload the entire object.
         * @param {number} [firstPart] The index of the first chunk to be uploaded.
         * @param {string} [uploadKey] The &#x60;&#x60;uploadKey&#x60;&#x60; of a previously-initiated upload, in order to request more chunk upload URLs for the same upload. If you do not specify a value, OSS will initiate a new upload entirely.
         * @param {number} [minutesExpiration] The time window (in minutes) the signed URL will remain usable. Acceptable values &#x3D; 1-60 minutes. Default &#x3D; 2 minutes.  **Tip:** Use the smallest possible time window to minimize exposure of the signed URL.
         * @param {boolean} [useAcceleration] &#x60;&#x60;true&#x60;&#x60; : (Default) Generates a faster S3 signed URL using Transfer Acceleration.  &#x60;&#x60;false&#x60;&#x60;: Generates a standard S3 signed URL.
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        signedS3Upload: (accessToken_1, bucketKey_1, objectKey_1, parts_1, firstPart_1, uploadKey_1, minutesExpiration_1, useAcceleration_1, ...args_1) => __awaiter(this, [accessToken_1, bucketKey_1, objectKey_1, parts_1, firstPart_1, uploadKey_1, minutesExpiration_1, useAcceleration_1, ...args_1], void 0, function* (accessToken, bucketKey, objectKey, parts, firstPart, uploadKey, minutesExpiration, useAcceleration, options = {}) {
            // verify required parameter 'bucketKey' is not null or undefined
            (0, common_1.assertParamExists)('signedS3Upload', 'bucketKey', bucketKey);
            // verify required parameter 'objectKey' is not null or undefined
            (0, common_1.assertParamExists)('signedS3Upload', 'objectKey', objectKey);
            const localVarPath = `/oss/v2/buckets/{bucketKey}/objects/{objectKey}/signeds3upload`
                .replace(`{${"bucketKey"}}`, encodeURIComponent(String(bucketKey)))
                .replace(`{${"objectKey"}}`, encodeURIComponent(String(objectKey)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'GET' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            if (parts !== undefined) {
                localVarQueryParameter['parts'] = parts;
            }
            if (firstPart !== undefined) {
                localVarQueryParameter['firstPart'] = firstPart;
            }
            if (uploadKey !== undefined) {
                localVarQueryParameter['uploadKey'] = uploadKey;
            }
            if (minutesExpiration !== undefined) {
                localVarQueryParameter['minutesExpiration'] = minutesExpiration;
            }
            if (useAcceleration !== undefined) {
                localVarQueryParameter['useAcceleration'] = useAcceleration;
            }
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Replaces an object that already exists on OSS, using an OSS signed URL.   The signed URL must fulfil the following conditions:  - The signed URL is valid (it has not expired as yet). - It was generated with ``write`` or ``readwrite`` for the ``access`` parameter.
         * @summary Replace Object Using Signed URL
         * @param {string} hash The ID component of the signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains &#x60;&#x60;hash&#x60;&#x60; as a URI parameter.
         * @param {number} contentLength The size of the data contained in the request body, in bytes.
         * @param {File} body The object to upload.
         * @param {string} [contentType] The MIME type of the object to upload; can be any type except \&#39;multipart/form-data\&#39;. This can be omitted, but we recommend adding it.
         * @param {string} [contentDisposition] The suggested file name to use when this object is downloaded as a file.
         * @param {Region} [xAdsRegion] Specifies where the bucket containing the object stored. Possible values are:  - &#x60;&#x60;US&#x60;&#x60; - (Default) Data center for the US region. - &#x60;&#x60;EMEA&#x60;&#x60; - Data center for the European Union, Middle East, and Africa. - &#x60;&#x60;APAC&#x60;&#x60; -  (Beta) Data center for Australia.  **Note:** Beta features are subject to change. Please do not use in production environments.
         * @param {string} [ifMatch] The current value of the &#x60;&#x60;sha1&#x60;&#x60; attribute of the object you want to replace. OSS checks the &#x60;&#x60;If-Match&#x60;&#x60; header against the &#x60;&#x60;sha1&#x60;&#x60; attribute of the object in OSS. If they match, OSS allows the object to be overwritten. Otherwise, it means that the object on OSS has been modified since you retrieved the &#x60;&#x60;sha1&#x60;&#x60; and the request fails.
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        uploadSignedResource: (accessToken_1, hash_1, contentLength_1, body_1, contentType_1, contentDisposition_1, xAdsRegion_1, ifMatch_1, ...args_1) => __awaiter(this, [accessToken_1, hash_1, contentLength_1, body_1, contentType_1, contentDisposition_1, xAdsRegion_1, ifMatch_1, ...args_1], void 0, function* (accessToken, hash, contentLength, body, contentType, contentDisposition, xAdsRegion, ifMatch, options = {}) {
            // verify required parameter 'hash' is not null or undefined
            (0, common_1.assertParamExists)('uploadSignedResource', 'hash', hash);
            // verify required parameter 'contentLength' is not null or undefined
            (0, common_1.assertParamExists)('uploadSignedResource', 'contentLength', contentLength);
            // verify required parameter 'body' is not null or undefined
            (0, common_1.assertParamExists)('uploadSignedResource', 'body', body);
            const localVarPath = `/oss/v2/signedresources/{hash}`
                .replace(`{${"hash"}}`, encodeURIComponent(String(hash)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'PUT' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            const localVarFormParams = new URLSearchParams();
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            if (contentType != null) {
                localVarHeaderParameter['Content-Type'] = String(contentType);
            }
            if (contentLength != null) {
                localVarHeaderParameter['Content-Length'] = typeof contentLength === 'string'
                    ? contentLength
                    : JSON.stringify(contentLength);
            }
            if (contentDisposition != null) {
                localVarHeaderParameter['Content-Disposition'] = String(contentDisposition);
            }
            if (xAdsRegion != null) {
                localVarHeaderParameter['x-ads-region'] = typeof xAdsRegion === 'string'
                    ? xAdsRegion
                    : JSON.stringify(xAdsRegion);
            }
            if (ifMatch != null) {
                localVarHeaderParameter['If-Match'] = String(ifMatch);
            }
            if (body !== undefined) {
                localVarFormParams.set('body', body);
            }
            localVarHeaderParameter['Content-Type'] = 'application/x-www-form-urlencoded';
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = localVarFormParams.toString();
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
        /**
         * Performs a resumable upload using an OSS signed URL. Use this operation to upload an object in chunks.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains the ``hash`` as a URI parameter.
         * @summary Upload Object Using Signed URL
         * @param {string} hash The ID component of the signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains &#x60;&#x60;hash&#x60;&#x60; as a URI parameter.
         * @param {string} contentRange The byte range to upload, specified in the form &#x60;&#x60;bytes&#x3D;&lt;START_BYTE&gt;-&lt;END_BYTE&gt;&#x60;&#x60;.
         * @param {string} sessionId An ID to uniquely identify the file upload session.
         * @param {File} body The chunk to upload.
         * @param {string} [contentType] The MIME type of the object to upload; can be any type except \&#39;multipart/form-data\&#39;. This can be omitted, but we recommend adding it.
         * @param {string} [contentDisposition] The suggested file name to use when this object is downloaded as a file.
         * @param {Region} [xAdsRegion] Specifies where the bucket containing the object stored. Possible values are:  - &#x60;&#x60;US&#x60;&#x60; - (Default) Data center for the US region. - &#x60;&#x60;EMEA&#x60;&#x60; - Data center for the European Union, Middle East, and Africa. - &#x60;&#x60;APAC&#x60;&#x60; -  (Beta) Data center for Australia.  **Note:** Beta features are subject to change. Please do not use in production environments.
         * @param accessToken bearer access token
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        uploadSignedResourcesChunk: (accessToken_1, hash_1, contentRange_1, sessionId_1, body_1, contentType_1, contentDisposition_1, xAdsRegion_1, ...args_1) => __awaiter(this, [accessToken_1, hash_1, contentRange_1, sessionId_1, body_1, contentType_1, contentDisposition_1, xAdsRegion_1, ...args_1], void 0, function* (accessToken, hash, contentRange, sessionId, body, contentType, contentDisposition, xAdsRegion, options = {}) {
            // verify required parameter 'hash' is not null or undefined
            (0, common_1.assertParamExists)('uploadSignedResourcesChunk', 'hash', hash);
            // verify required parameter 'contentRange' is not null or undefined
            (0, common_1.assertParamExists)('uploadSignedResourcesChunk', 'contentRange', contentRange);
            // verify required parameter 'sessionId' is not null or undefined
            (0, common_1.assertParamExists)('uploadSignedResourcesChunk', 'sessionId', sessionId);
            // verify required parameter 'body' is not null or undefined
            (0, common_1.assertParamExists)('uploadSignedResourcesChunk', 'body', body);
            const localVarPath = `/oss/v2/signedresources/{hash}/resumable`
                .replace(`{${"hash"}}`, encodeURIComponent(String(hash)));
            const localVarUrlObj = new URL(localVarPath, apsConfiguration.baseAddress);
            let baseOptions;
            if (apsConfiguration) {
                baseOptions = apsConfiguration.baseOptions;
            }
            const localVarRequestOptions = Object.assign(Object.assign({ method: 'PUT' }, baseOptions), options);
            const localVarHeaderParameter = {};
            const localVarQueryParameter = {};
            const localVarFormParams = new URLSearchParams();
            yield (0, common_1.setBearerAuthToObject)(localVarHeaderParameter, accessToken);
            if (contentType != null) {
                localVarHeaderParameter['Content-Type'] = String(contentType);
            }
            if (contentRange != null) {
                localVarHeaderParameter['Content-Range'] = String(contentRange);
            }
            if (contentDisposition != null) {
                localVarHeaderParameter['Content-Disposition'] = String(contentDisposition);
            }
            if (xAdsRegion != null) {
                localVarHeaderParameter['x-ads-region'] = typeof xAdsRegion === 'string'
                    ? xAdsRegion
                    : JSON.stringify(xAdsRegion);
            }
            if (sessionId != null) {
                localVarHeaderParameter['Session-Id'] = String(sessionId);
            }
            if (body !== undefined) {
                localVarFormParams.set('body', body);
            }
            localVarHeaderParameter['Content-Type'] = 'application/x-www-form-urlencoded';
            localVarHeaderParameter['User-Agent'] = 'APS SDK/OSS/TypeScript/1.0.0';
            (0, common_1.setSearchParams)(localVarUrlObj, localVarQueryParameter);
            let headersFromBaseOptions = baseOptions && baseOptions.headers ? baseOptions.headers : {};
            localVarRequestOptions.headers = Object.assign(Object.assign(Object.assign({}, localVarHeaderParameter), headersFromBaseOptions), options.headers);
            localVarRequestOptions.data = localVarFormParams.toString();
            return {
                url: (0, common_1.toPathString)(localVarUrlObj),
                options: localVarRequestOptions,
            };
        }),
    };
};
exports.ObjectsApiAxiosParamCreator = ObjectsApiAxiosParamCreator;
/**
 * ObjectsApi - functional programming interface
 * @export
 */
const ObjectsApiFp = function (sdkManager) {
    const localVarAxiosParamCreator = (0, exports.ObjectsApiAxiosParamCreator)(sdkManager.apsConfiguration);
    return {
        /**
         * Requests OSS to start reconstituting the set of objects that were uploaded using signed S3 upload URLs. You must call this operation only after all the objects have been uploaded.   You can specify up to 25 objects in this operation.
         * @summary Complete Batch Upload to S3 Signed URLs
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {BatchcompleteuploadObject} [requests] An array of objects, each of which represents an upload to complete.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        batchCompleteUpload(accessToken, bucketKey, requests, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.batchCompleteUpload(accessToken, bucketKey, requests, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Creates and returns signed URLs to download a set of objects directly from S3. These signed URLs expire in 2 minutes by default, but you can change this duration if needed.  You must start download the objects before the signed URLs expire. The download itself can take longer.  Only the application that owns the bucket can call this operation. All other applications that call this operation will receive a \"403 Forbidden\" error.
         * @summary Batch Generate Signed S3 Download URLs
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {Batchsigneds3downloadObject} requests An array of objects representing each request for a signed download URL.
         * @param {boolean} [publicResourceFallback] Specifies how to return the signed URLs, in case the object was uploaded in chunks, and assembling of chunks is not yet complete.  - &#x60;&#x60;true&#x60;&#x60; : Return a single signed OSS URL. - &#x60;&#x60;false&#x60;&#x60; : (Default) Return multiple signed S3 URLs, where each URL points to a chunk.
         * @param {number} [minutesExpiration] The time window (in minutes) the signed URL will remain usable. Acceptable values &#x3D; 1-60 minutes. Default &#x3D; 2 minutes.  **Tip:** Use the smallest possible time window to minimize exposure of the signed URL.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        batchSignedS3Download(accessToken, bucketKey, requests, publicResourceFallback, minutesExpiration, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.batchSignedS3Download(accessToken, bucketKey, requests, publicResourceFallback, minutesExpiration, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Creates and returns signed URLs to upload a set of objects directly to S3. These signed URLs expire in 2 minutes by default, but you can change this duration if needed.  You must start uploading the objects before the signed URLs expire. The upload  itself can take longer.  Only the application that owns the bucket can call this operation. All other applications that call this operation will receive a \"403 Forbidden\" error.  If required, you can request an array of signed URLs for each object, which lets you upload the objects in chunks. Once you upload all chunks you must call the [Complete Batch Upload to S3 Signed URL](/en/docs/data/v2/reference/http/buckets-:bucketKey-objects-:objectKey-batchcompleteupload-POST/) operation to indicate completion. This instructs OSS to assemble the chunks and reconstitute the object on OSS. You must call this operation even if you requested a single signed URL for an object.  If an upload fails after the validity period of a signed URL has elapsed, you can call this operation again to obtain fresh signed URLs. However, you must use the same ``uploadKey`` that was returned when you originally called this operation.
         * @summary Batch Generate Signed S3 Upload URLs
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {boolean} [useAcceleration] &#x60;&#x60;true&#x60;&#x60; : (Default) Generates a faster S3 signed URL using Transfer Acceleration.  &#x60;&#x60;false&#x60;&#x60;: Generates a standard S3 signed URL.
         * @param {number} [minutesExpiration] The time window (in minutes) the signed URL will remain usable. Acceptable values &#x3D; 1-60 minutes. Default &#x3D; 2 minutes.  **Tip:** Use the smallest possible time window to minimize exposure of the signed URL.
         * @param {Batchsigneds3uploadObject} [requests] An array of objects representing each request for a signed upload URL.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        batchSignedS3Upload(accessToken, bucketKey, useAcceleration, minutesExpiration, requests, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.batchSignedS3Upload(accessToken, bucketKey, useAcceleration, minutesExpiration, requests, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Requests OSS to assemble and reconstitute the object that was uploaded using signed S3 upload URL. You must call this operation only after all parts/chunks of the object has been uploaded.
         * @summary Complete Upload to S3 Signed URL
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {string} contentType Must be &#x60;&#x60;application/json&#x60;&#x60;.
         * @param {Completes3uploadBody} body
         * @param {string} [xAdsMetaContentType] The Content-Type value for the uploaded object to record within OSS.
         * @param {string} [xAdsMetaContentDisposition] The Content-Disposition value for the uploaded object to record within OSS.
         * @param {string} [xAdsMetaContentEncoding] The Content-Encoding value for the uploaded object to record within OSS.
         * @param {string} [xAdsMetaCacheControl] The Cache-Control value for the uploaded object to record within OSS.
         * @param {string} [xAdsUserDefinedMetadata] Custom metadata to be stored with the object, which can be retrieved later on download or when retrieving object details. Must be a JSON object that is less than 100 bytes.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        completeSignedS3Upload(accessToken, bucketKey, objectKey, contentType, body, xAdsMetaContentType, xAdsMetaContentDisposition, xAdsMetaContentEncoding, xAdsMetaCacheControl, xAdsUserDefinedMetadata, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.completeSignedS3Upload(accessToken, bucketKey, objectKey, contentType, body, xAdsMetaContentType, xAdsMetaContentDisposition, xAdsMetaContentEncoding, xAdsMetaCacheControl, xAdsUserDefinedMetadata, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Creates a copy of an object within the bucket.
         * @summary Copy Object
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {string} newObjName A URL-encoded human friendly name to identify the copied object.
         * @param {string} [xAdsAcmNamespace] This header is used to let the OSS Api Proxy know if ACM is used to authorize access to the given object. If this authorization is used by your service, then you must provide the name of the namespace you want to validate access control policies against.
         * @param {string} [xAdsAcmCheckGroups] Informs the OSS Api Proxy know if your service requires ACM authorization to also validate against Oxygen groups. If so, you must pass this header with a value of &#x60;&#x60;true&#x60;&#x60;. Otherwise, the assumption is that checking authorization against Oxygen groups is not required.
         * @param {string} [xAdsAcmGroups] Use this header to pass the Oxygen groups you want the OSS Api Proxy to use for group validation for the given user in the OAuth2 token.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        copyTo(accessToken, bucketKey, objectKey, newObjName, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.copyTo(accessToken, bucketKey, objectKey, newObjName, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Generates a signed URL that can be used to download or upload an object within the specified expiration time. If the object the signed URL points to is deleted or expires before the signed URL expires, the signed URL will no longer be valid.  In addition to this operation that generates OSS signed URLs, OSS provides operations to generate S3 signed URLs. S3 signed URLs allow direct upload/download from S3 but are restricted to bucket owners. OSS signed URLs also allow upload/download and can be configured for access by other applications, making them suitable for sharing objects across applications.  Only the application that owns the bucket containing the object can call this operation.
         * @summary Generate OSS Signed URL
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {Access} [access]
         * @param {boolean} [useCdn] &#x60;&#x60;true&#x60;&#x60; : Returns a Cloudfront URL to the object instead of an Autodesk URL (one that points to a location on https://developer.api.autodesk.com). Applications can interact with the Cloudfront URL exactly like with any classic public resource in OSS. They can use GET requests to download objects or PUT requests to upload objects.  &#x60;&#x60;false&#x60;&#x60; : (Default) Returns an Autodesk URL (one that points to a location on https://developer.api.autodesk.com) to the object.
         * @param {CreateSignedResource} [createSignedResource]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        createSignedResource(accessToken, bucketKey, objectKey, access, useCdn, createSignedResource, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.createSignedResource(accessToken, bucketKey, objectKey, access, useCdn, createSignedResource, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Deletes an object from the bucket.
         * @summary Delete Object
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {string} [xAdsAcmNamespace] This header is used to let the OSS Api Proxy know if ACM is used to authorize access to the given object. If this authorization is used by your service, then you must provide the name of the namespace you want to validate access control policies against.
         * @param {string} [xAdsAcmCheckGroups] Informs the OSS API Proxy know if your service requires ACM authorization to also validate against Oxygen groups. If so, you must pass this header with a value of &#x60;&#x60;true&#x60;&#x60;. Otherwise, the assumption is that checking authorization against Oxygen groups is not required.
         * @param {string} [xAdsAcmGroups] Use this header to pass the Oxygen groups you want the OSS Api Proxy to use for group validation for the given user in the OAuth2 token.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteObject(accessToken, bucketKey, objectKey, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.deleteObject(accessToken, bucketKey, objectKey, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Delete an object using an OSS signed URL to access it.  Only applications that own the bucket containing the object can call this operation.
         * @summary Delete Object Using Signed URL
         * @param {string} hash The ID component of the signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains &#x60;&#x60;hash&#x60;&#x60; as a URI parameter.
         * @param {Region} [xAdsRegion] Specifies where the bucket containing the object stored. Possible values are:  - &#x60;&#x60;US&#x60;&#x60; - (Default) Data center for the US region. - &#x60;&#x60;EMEA&#x60;&#x60; - Data center for the European Union, Middle East, and Africa. - &#x60;&#x60;APAC&#x60;&#x60; -  (Beta) Data center for Australia.  **Note:** Beta features are subject to change. Please do not use in production environments.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        deleteSignedResource(accessToken, hash, xAdsRegion, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.deleteSignedResource(accessToken, hash, xAdsRegion, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Returns detailed information about the specified object.
         * @summary Get Object Details
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {string} [ifModifiedSince] A timestamp in the HTTP date format (Mon, DD Month YYYY HH:MM:SS GMT). The requested data is returned only if the object has been modified since the specified timestamp. If not, a 304 (Not Modified) HTTP status is returned.
         * @param {string} [xAdsAcmNamespace] This header is used to let the OSS Api Proxy know if ACM is used to authorize access to the given object. If this authorization is used by your service, then you must provide the name of the namespace you want to validate access control policies against.
         * @param {string} [xAdsAcmCheckGroups] Informs the OSS Api Proxy know if your service requires ACM authorization to also validate against Oxygen groups. If so, you must pass this header with a value of &#x60;&#x60;true&#x60;&#x60;. Otherwise, the assumption is that checking authorization against Oxygen groups is not required.
         * @param {string} [xAdsAcmGroups] Use this header to pass the Oxygen groups you want the OSS Api Proxy to use for group validation for the given user in the OAuth2 token.
         * @param {With} [_with]
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getObjectDetails(accessToken, bucketKey, objectKey, ifModifiedSince, _with, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getObjectDetails(accessToken, bucketKey, objectKey, ifModifiedSince, _with, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Returns a list objects in the specified bucket.   Only the application that owns the bucket can call this operation. All other applications that call this operation will receive a \"403 Forbidden\" error.
         * @summary List Objects
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {number} [limit] The number of items you want per page. Acceptable values &#x3D; 1-100. Default &#x3D; 10.
         * @param {string} [beginsWith] Filters the results by the value you specify. Only the objects with their &#x60;&#x60;objectKey&#x60;&#x60; beginning with the specified string are returned.
         * @param {string} [startAt] The ID of the last item that was returned in the previous result set.  It enables the system to return subsequent items starting from the next one after the specified ID.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getObjects(accessToken, bucketKey, limit, beginsWith, startAt, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getObjects(accessToken, bucketKey, limit, beginsWith, startAt, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Downloads an object using an OSS signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/)  contains the ``hash`` URI parameter as well.
         * @summary Download Object Using Signed URL
         * @param {string} hash The ID component of the signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains &#x60;&#x60;hash&#x60;&#x60; as a URI parameter.
         * @param {string} [range] The byte range to download, specified in the form &#x60;&#x60;bytes&#x3D;&lt;START_BYTE&gt;-&lt;END_BYTE&gt;&#x60;&#x60;.
         * @param {string} [ifNoneMatch] The last known ETag value of the object. OSS returns the requested data  only if the &#x60;&#x60;If-None-Match&#x60;&#x60; header differs from the ETag value of the object on OSS, which indicates that the object on OSS is newer. If not, it returns a 304 \&quot;Not Modified\&quot; HTTP status.
         * @param {string} [ifModifiedSince] A timestamp in the HTTP date format (Mon, DD Month YYYY HH:MM:SS GMT). The requested data is returned only if the object has been modified since the specified timestamp. If not, a 304 (Not Modified) HTTP status is returned.
         * @param {string} [acceptEncoding] The compression format your prefer to receive the data. Possible values are:  - &#x60;&#x60;gzip&#x60;&#x60; - Use the gzip format  **Note:** You cannot use &#x60;&#x60;Accept-Encoding:gzip&#x60;&#x60; with a Range header containing an end byte range. OSS will not honor the End byte range if &#x60;&#x60;Accept-Encoding: gzip&#x60;&#x60; header is used.
         * @param {Region} [region] Specifies where the bucket containing the object stored. Possible values are:  - &#x60;&#x60;US&#x60;&#x60; - (Default) Data center for the US region. - &#x60;&#x60;EMEA&#x60;&#x60; - Data center for the European Union, Middle East, and Africa. - &#x60;&#x60;APAC&#x60;&#x60; -  (Beta) Data center for Australia.  **Note:** Beta features are subject to change. Please do not use in production environments.
         * @param {string} [responseContentDisposition] The value of the Content-Disposition header you want to receive when you download the object using the signed URL. If you do not specify a value, the Content-Disposition header defaults to the value stored with OSS.
         * @param {string} [responseContentType] The value of the Content-Type header you want to receive when you download the object using the signed URL. If you do not specify a value, the Content-Type header defaults to the value stored with OSS.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        getSignedResource(accessToken, hash, range, ifNoneMatch, ifModifiedSince, acceptEncoding, region, responseContentDisposition, responseContentType, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.getSignedResource(accessToken, hash, range, ifNoneMatch, ifModifiedSince, acceptEncoding, region, responseContentDisposition, responseContentType, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Gets a signed URL to download an object directly from S3, bypassing OSS servers. This signed URL expires in 2 minutes by default, but you can change this duration if needed.  You must start the download before the signed URL expires. The download itself can take longer. If the download fails after the validity period of the signed URL has elapsed, you can call this operation again to obtain a fresh signed URL.  Only applications that have read access to the object can call this operation.     You can use range headers with the signed download URL to download the object in chunks. This ability lets you download chunks in parallel, which can result in faster downloads.  If the object you want to download was uploaded in chunks and is still assembling on OSS, you will receive multiple S3 URLs instead of just one. Each URL will point to a chunk. If you prefer to receive a single URL, set the ``public-resource-fallback`` query parameter to ``true``. This setting will make OSS fallback to returning a single signed OSS URL, if assembling is still in progress.   In addition to this operation that generates S3 signed URLs, OSS provides an operation to generate OSS signed URLs. S3 signed URLs allow direct upload/download from S3 but are restricted to bucket owners. OSS signed URLs also allow upload/download and can be configured for access by other applications, making them suitable for sharing objects across applications.
         * @summary Generate Signed S3 Download URL
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {string} [ifNoneMatch] The last known ETag value of the object. OSS returns the signed URL only if the &#x60;&#x60;If-None-Match&#x60;&#x60; header differs from the ETag value of the object on S3. If not, it returns a 304 \&quot;Not Modified\&quot; HTTP status.
         * @param {string} [ifModifiedSince] A timestamp in the HTTP date format (Mon, DD Month YYYY HH:MM:SS GMT). The signed URL is returned only if the object has been modified since the specified timestamp. If not, a 304 (Not Modified) HTTP status is returned.
         * @param {string} [xAdsAcmScopes] Optional OSS-compliant scope reference to increase bucket search performance
         * @param {string} [responseContentType] The value of the Content-Type header you want to receive when you download the object using the signed URL. If you do not specify a value, the Content-Type header defaults to the value stored with OSS.
         * @param {string} [responseContentDisposition] The value of the Content-Disposition header you want to receive when you download the object using the signed URL. If you do not specify a value, the Content-Disposition header defaults to the value stored with OSS.
         * @param {string} [responseCacheControl] The value of the Cache-Control header you want to receive when you download the object using the signed URL. If you do not specify a value, the Cache-Control header defaults to the value stored with OSS.
         * @param {boolean} [publicResourceFallback] Specifies how to return the signed URLs, in case the object was uploaded in chunks, and assembling of chunks is not yet complete.  - &#x60;&#x60;true&#x60;&#x60; : Return a single signed OSS URL. - &#x60;&#x60;false&#x60;&#x60; : (Default) Return multiple signed S3 URLs, where each URL points to a chunk.
         * @param {number} [minutesExpiration] The time window (in minutes) the signed URL will remain usable. Acceptable values &#x3D; 1-60 minutes. Default &#x3D; 2 minutes.  **Tip:** Use the smallest possible time window to minimize exposure of the signed URL.
         * @param {boolean} [useCdn] &#x60;&#x60;true&#x60;&#x60; : Returns a URL that points to a CloudFront edge location.  &#x60;&#x60;false&#x60;&#x60; : (Default) Returns a URL that points directly to the S3 object.
         * @param {boolean} [redirect] Generates a HTTP redirection response (Temporary Redirect 307​) using the generated URL.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        signedS3Download(accessToken, bucketKey, objectKey, ifNoneMatch, ifModifiedSince, responseContentType, responseContentDisposition, responseCacheControl, publicResourceFallback, minutesExpiration, useCdn, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.signedS3Download(accessToken, bucketKey, objectKey, ifNoneMatch, ifModifiedSince, responseContentType, responseContentDisposition, responseCacheControl, publicResourceFallback, minutesExpiration, useCdn, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Gets a signed URL to upload an object directly to S3, bypassing OSS servers. You can also request an array of signed URLs which lets you upload an object in chunks.  This signed URL expires in 2 minutes by default, but you can change this duration if needed.  You must start the upload before the signed URL expires. The upload itself can take longer. If the upload fails after the validity period of the signed URL has elapsed, you can call this operation again to obtain a fresh signed URL (or an array of signed URLs as the case may be). However, you must use the same ``uploadKey`` that was returned when you originally called this operation.   Only applications that own the bucket can call this operation.  **Note:** Once you upload all chunks you must call the [Complete Upload to S3 Signed URL](/en/docs/data/v2/reference/http/buckets-:bucketKey-objects-:objectKey-signeds3upload-POST/) operation to indicate completion. This instructs OSS to assemble the chunks and reconstitute the object on OSS. You must call this operation even when using a single signed URL.   In addition to this operation that generates S3 signed URLs, OSS provides an operation to generate OSS signed URLs. S3 signed URLs allow direct upload/download from S3 but are restricted to bucket owners. OSS signed URLs also allow upload/download and can be configured for access by other applications, making them suitable for sharing objects across applications.
         * @summary Generate Signed S3 Upload URL
         * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
         * @param {string} objectKey The URL-encoded human friendly name of the object.
         * @param {string} [xAdsAcmScopes] Optional OSS-compliant scope reference to increase bucket search performance
         * @param {number} [parts] The number of parts you intend to chunk the object for uploading. OSS will return that many signed URLs, one URL for each chunk. If you do not specify a value you\&#39;ll get only one URL to upload the entire object.
         * @param {number} [firstPart] The index of the first chunk to be uploaded.
         * @param {string} [uploadKey] The &#x60;&#x60;uploadKey&#x60;&#x60; of a previously-initiated upload, in order to request more chunk upload URLs for the same upload. If you do not specify a value, OSS will initiate a new upload entirely.
         * @param {number} [minutesExpiration] The time window (in minutes) the signed URL will remain usable. Acceptable values &#x3D; 1-60 minutes. Default &#x3D; 2 minutes.  **Tip:** Use the smallest possible time window to minimize exposure of the signed URL.
         * @param {boolean} [useAcceleration] &#x60;&#x60;true&#x60;&#x60; : (Default) Generates a faster S3 signed URL using Transfer Acceleration.  &#x60;&#x60;false&#x60;&#x60;: Generates a standard S3 signed URL.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        signedS3Upload(accessToken, bucketKey, objectKey, parts, firstPart, uploadKey, minutesExpiration, useAcceleration, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.signedS3Upload(accessToken, bucketKey, objectKey, parts, firstPart, uploadKey, minutesExpiration, useAcceleration, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Replaces an object that already exists on OSS, using an OSS signed URL.   The signed URL must fulfil the following conditions:  - The signed URL is valid (it has not expired as yet). - It was generated with ``write`` or ``readwrite`` for the ``access`` parameter.
         * @summary Replace Object Using Signed URL
         * @param {string} hash The ID component of the signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains &#x60;&#x60;hash&#x60;&#x60; as a URI parameter.
         * @param {number} contentLength The size of the data contained in the request body, in bytes.
         * @param {File} body The object to upload.
         * @param {string} [contentType] The MIME type of the object to upload; can be any type except \&#39;multipart/form-data\&#39;. This can be omitted, but we recommend adding it.
         * @param {string} [contentDisposition] The suggested file name to use when this object is downloaded as a file.
         * @param {Region} [xAdsRegion] Specifies where the bucket containing the object stored. Possible values are:  - &#x60;&#x60;US&#x60;&#x60; - (Default) Data center for the US region. - &#x60;&#x60;EMEA&#x60;&#x60; - Data center for the European Union, Middle East, and Africa. - &#x60;&#x60;APAC&#x60;&#x60; -  (Beta) Data center for Australia.  **Note:** Beta features are subject to change. Please do not use in production environments.
         * @param {string} [ifMatch] The current value of the &#x60;&#x60;sha1&#x60;&#x60; attribute of the object you want to replace. OSS checks the &#x60;&#x60;If-Match&#x60;&#x60; header against the &#x60;&#x60;sha1&#x60;&#x60; attribute of the object in OSS. If they match, OSS allows the object to be overwritten. Otherwise, it means that the object on OSS has been modified since you retrieved the &#x60;&#x60;sha1&#x60;&#x60; and the request fails.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        uploadSignedResource(accessToken, hash, contentLength, body, contentType, contentDisposition, xAdsRegion, ifMatch, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.uploadSignedResource(accessToken, hash, contentLength, body, contentType, contentDisposition, xAdsRegion, ifMatch, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
        /**
         * Performs a resumable upload using an OSS signed URL. Use this operation to upload an object in chunks.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains the ``hash`` as a URI parameter.
         * @summary Upload Object Using Signed URL
         * @param {string} hash The ID component of the signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains &#x60;&#x60;hash&#x60;&#x60; as a URI parameter.
         * @param {string} contentRange The byte range to upload, specified in the form &#x60;&#x60;bytes&#x3D;&lt;START_BYTE&gt;-&lt;END_BYTE&gt;&#x60;&#x60;.
         * @param {string} sessionId An ID to uniquely identify the file upload session.
         * @param {File} body The chunk to upload.
         * @param {string} [contentType] The MIME type of the object to upload; can be any type except \&#39;multipart/form-data\&#39;. This can be omitted, but we recommend adding it.
         * @param {string} [contentDisposition] The suggested file name to use when this object is downloaded as a file.
         * @param {Region} [xAdsRegion] Specifies where the bucket containing the object stored. Possible values are:  - &#x60;&#x60;US&#x60;&#x60; - (Default) Data center for the US region. - &#x60;&#x60;EMEA&#x60;&#x60; - Data center for the European Union, Middle East, and Africa. - &#x60;&#x60;APAC&#x60;&#x60; -  (Beta) Data center for Australia.  **Note:** Beta features are subject to change. Please do not use in production environments.
         * @param {*} [options] Override http request option.
         * @throws {RequiredError}
         */
        uploadSignedResourcesChunk(accessToken, hash, contentRange, sessionId, body, contentType, contentDisposition, xAdsRegion, options) {
            return __awaiter(this, void 0, void 0, function* () {
                const localVarAxiosArgs = yield localVarAxiosParamCreator.uploadSignedResourcesChunk(accessToken, hash, contentRange, sessionId, body, contentType, contentDisposition, xAdsRegion, options);
                return (0, common_1.createRequestFunction)(localVarAxiosArgs, sdkManager);
            });
        },
    };
};
exports.ObjectsApiFp = ObjectsApiFp;
/**
 * ObjectsApi - object-oriented interface
 * @export
 * @class ObjectsApi
 * @extends {BaseApi}
 */
class ObjectsApi extends base_1.BaseApi {
    constructor() {
        super(...arguments);
        this.logger = this.sdkManager.logger;
    }
    /**
     * Requests OSS to start reconstituting the set of objects that were uploaded using signed S3 upload URLs. You must call this operation only after all the objects have been uploaded.   You can specify up to 25 objects in this operation.
     * @summary Complete Batch Upload to S3 Signed URLs
     * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
     * @param {BatchcompleteuploadObject} [requests] An array of objects, each of which represents an upload to complete.
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    batchCompleteUpload(accessToken, bucketKey, requests, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into batchCompleteUpload ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).batchCompleteUpload(accessToken, bucketKey, requests, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`batchCompleteUpload Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`batchCompleteUpload Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`batchCompleteUpload Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`batchCompleteUpload Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`batchCompleteUpload Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Creates and returns signed URLs to download a set of objects directly from S3. These signed URLs expire in 2 minutes by default, but you can change this duration if needed.  You must start download the objects before the signed URLs expire. The download itself can take longer.  Only the application that owns the bucket can call this operation. All other applications that call this operation will receive a \"403 Forbidden\" error.
     * @summary Batch Generate Signed S3 Download URLs
     * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
     * @param {Batchsigneds3downloadObject} requests An array of objects representing each request for a signed download URL.
     * @param {boolean} [publicResourceFallback] Specifies how to return the signed URLs, in case the object was uploaded in chunks, and assembling of chunks is not yet complete.  - &#x60;&#x60;true&#x60;&#x60; : Return a single signed OSS URL. - &#x60;&#x60;false&#x60;&#x60; : (Default) Return multiple signed S3 URLs, where each URL points to a chunk.
     * @param {number} [minutesExpiration] The time window (in minutes) the signed URL will remain usable. Acceptable values &#x3D; 1-60 minutes. Default &#x3D; 2 minutes.  **Tip:** Use the smallest possible time window to minimize exposure of the signed URL.
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    batchSignedS3Download(accessToken, bucketKey, requests, publicResourceFallback, minutesExpiration, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into batchSignedS3Download ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).batchSignedS3Download(accessToken, bucketKey, requests, publicResourceFallback, minutesExpiration, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`batchSignedS3Download Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`batchSignedS3Download Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`batchSignedS3Download Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`batchSignedS3Download Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`batchSignedS3Download Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Creates and returns signed URLs to upload a set of objects directly to S3. These signed URLs expire in 2 minutes by default, but you can change this duration if needed.  You must start uploading the objects before the signed URLs expire. The upload  itself can take longer.  Only the application that owns the bucket can call this operation. All other applications that call this operation will receive a \"403 Forbidden\" error.  If required, you can request an array of signed URLs for each object, which lets you upload the objects in chunks. Once you upload all chunks you must call the [Complete Batch Upload to S3 Signed URL](/en/docs/data/v2/reference/http/buckets-:bucketKey-objects-:objectKey-batchcompleteupload-POST/) operation to indicate completion. This instructs OSS to assemble the chunks and reconstitute the object on OSS. You must call this operation even if you requested a single signed URL for an object.  If an upload fails after the validity period of a signed URL has elapsed, you can call this operation again to obtain fresh signed URLs. However, you must use the same ``uploadKey`` that was returned when you originally called this operation.
     * @summary Batch Generate Signed S3 Upload URLs
     * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
     * @param {boolean} [useAcceleration] &#x60;&#x60;true&#x60;&#x60; : (Default) Generates a faster S3 signed URL using Transfer Acceleration.  &#x60;&#x60;false&#x60;&#x60;: Generates a standard S3 signed URL.
     * @param {number} [minutesExpiration] The time window (in minutes) the signed URL will remain usable. Acceptable values &#x3D; 1-60 minutes. Default &#x3D; 2 minutes.  **Tip:** Use the smallest possible time window to minimize exposure of the signed URL.
     * @param {Batchsigneds3uploadObject} [requests] An array of objects representing each request for a signed upload URL.
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    batchSignedS3Upload(accessToken, bucketKey, useAcceleration, minutesExpiration, requests, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into batchSignedS3Upload ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).batchSignedS3Upload(accessToken, bucketKey, useAcceleration, minutesExpiration, requests, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`batchSignedS3Upload Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`batchSignedS3Upload Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`batchSignedS3Upload Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`batchSignedS3Upload Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`batchSignedS3Upload Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Requests OSS to assemble and reconstitute the object that was uploaded using signed S3 upload URL. You must call this operation only after all parts/chunks of the object has been uploaded.
     * @summary Complete Upload to S3 Signed URL
     * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
     * @param {string} objectKey The URL-encoded human friendly name of the object.
     * @param {string} contentType Must be &#x60;&#x60;application/json&#x60;&#x60;.
     * @param {Completes3uploadBody} body
     * @param {string} [xAdsMetaContentType] The Content-Type value for the uploaded object to record within OSS.
     * @param {string} [xAdsMetaContentDisposition] The Content-Disposition value for the uploaded object to record within OSS.
     * @param {string} [xAdsMetaContentEncoding] The Content-Encoding value for the uploaded object to record within OSS.
     * @param {string} [xAdsMetaCacheControl] The Cache-Control value for the uploaded object to record within OSS.
     * @param {string} [xAdsUserDefinedMetadata] Custom metadata to be stored with the object, which can be retrieved later on download or when retrieving object details. Must be a JSON object that is less than 100 bytes.
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    completeSignedS3Upload(accessToken, bucketKey, objectKey, contentType, body, xAdsMetaContentType, xAdsMetaContentDisposition, xAdsMetaContentEncoding, xAdsMetaCacheControl, xAdsUserDefinedMetadata, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into completeSignedS3Upload ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).completeSignedS3Upload(accessToken, bucketKey, objectKey, contentType, body, xAdsMetaContentType, xAdsMetaContentDisposition, xAdsMetaContentEncoding, xAdsMetaCacheControl, xAdsUserDefinedMetadata, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`completeSignedS3Upload Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`completeSignedS3Upload Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`completeSignedS3Upload Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`completeSignedS3Upload Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`completeSignedS3Upload Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Creates a copy of an object within the bucket.
     * @summary Copy Object
     * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
     * @param {string} objectKey The URL-encoded human friendly name of the object.
     * @param {string} newObjName A URL-encoded human friendly name to identify the copied object.
     * @param {string} [xAdsAcmNamespace] This header is used to let the OSS Api Proxy know if ACM is used to authorize access to the given object. If this authorization is used by your service, then you must provide the name of the namespace you want to validate access control policies against.
     * @param {string} [xAdsAcmCheckGroups] Informs the OSS Api Proxy know if your service requires ACM authorization to also validate against Oxygen groups. If so, you must pass this header with a value of &#x60;&#x60;true&#x60;&#x60;. Otherwise, the assumption is that checking authorization against Oxygen groups is not required.
     * @param {string} [xAdsAcmGroups] Use this header to pass the Oxygen groups you want the OSS Api Proxy to use for group validation for the given user in the OAuth2 token.
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    copyTo(accessToken, bucketKey, objectKey, newObjName, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into copyTo ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).copyTo(accessToken, bucketKey, objectKey, newObjName, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`copyTo Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`copyTo Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`copyTo Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`copyTo Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`copyTo Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Generates a signed URL that can be used to download or upload an object within the specified expiration time. If the object the signed URL points to is deleted or expires before the signed URL expires, the signed URL will no longer be valid.  In addition to this operation that generates OSS signed URLs, OSS provides operations to generate S3 signed URLs. S3 signed URLs allow direct upload/download from S3 but are restricted to bucket owners. OSS signed URLs also allow upload/download and can be configured for access by other applications, making them suitable for sharing objects across applications.  Only the application that owns the bucket containing the object can call this operation.
     * @summary Generate OSS Signed URL
     * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
     * @param {string} objectKey The URL-encoded human friendly name of the object.
     * @param {Access} [access]
     * @param {boolean} [useCdn] &#x60;&#x60;true&#x60;&#x60; : Returns a Cloudfront URL to the object instead of an Autodesk URL (one that points to a location on https://developer.api.autodesk.com). Applications can interact with the Cloudfront URL exactly like with any classic public resource in OSS. They can use GET requests to download objects or PUT requests to upload objects.  &#x60;&#x60;false&#x60;&#x60; : (Default) Returns an Autodesk URL (one that points to a location on https://developer.api.autodesk.com) to the object.
     * @param {CreateSignedResource} [createSignedResource]
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    createSignedResource(accessToken, bucketKey, objectKey, access, useCdn, createSignedResource, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into createSignedResource ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).createSignedResource(accessToken, bucketKey, objectKey, access, useCdn, createSignedResource, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`createSignedResource Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`createSignedResource Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`createSignedResource Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`createSignedResource Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`createSignedResource Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Deletes an object from the bucket.
     * @summary Delete Object
     * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
     * @param {string} objectKey The URL-encoded human friendly name of the object.
     * @param {string} [xAdsAcmNamespace] This header is used to let the OSS Api Proxy know if ACM is used to authorize access to the given object. If this authorization is used by your service, then you must provide the name of the namespace you want to validate access control policies against.
     * @param {string} [xAdsAcmCheckGroups] Informs the OSS API Proxy know if your service requires ACM authorization to also validate against Oxygen groups. If so, you must pass this header with a value of &#x60;&#x60;true&#x60;&#x60;. Otherwise, the assumption is that checking authorization against Oxygen groups is not required.
     * @param {string} [xAdsAcmGroups] Use this header to pass the Oxygen groups you want the OSS Api Proxy to use for group validation for the given user in the OAuth2 token.
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    deleteObject(accessToken, bucketKey, objectKey, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into deleteObject ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).deleteObject(accessToken, bucketKey, objectKey, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`deleteObject Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`deleteObject Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`deleteObject Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`deleteObject Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`deleteObject Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Delete an object using an OSS signed URL to access it.  Only applications that own the bucket containing the object can call this operation.
     * @summary Delete Object Using Signed URL
     * @param {string} hash The ID component of the signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains &#x60;&#x60;hash&#x60;&#x60; as a URI parameter.
     * @param {Region} [xAdsRegion] Specifies where the bucket containing the object stored. Possible values are:  - &#x60;&#x60;US&#x60;&#x60; - (Default) Data center for the US region. - &#x60;&#x60;EMEA&#x60;&#x60; - Data center for the European Union, Middle East, and Africa. - &#x60;&#x60;APAC&#x60;&#x60; -  (Beta) Data center for Australia.  **Note:** Beta features are subject to change. Please do not use in production environments.
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    deleteSignedResource(accessToken, hash, xAdsRegion, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into deleteSignedResource ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).deleteSignedResource(accessToken, hash, xAdsRegion, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`deleteSignedResource Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`deleteSignedResource Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`deleteSignedResource Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`deleteSignedResource Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`deleteSignedResource Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Returns detailed information about the specified object.
     * @summary Get Object Details
     * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
     * @param {string} objectKey The URL-encoded human friendly name of the object.
     * @param {string} [ifModifiedSince] A timestamp in the HTTP date format (Mon, DD Month YYYY HH:MM:SS GMT). The requested data is returned only if the object has been modified since the specified timestamp. If not, a 304 (Not Modified) HTTP status is returned.
     * @param {string} [xAdsAcmNamespace] This header is used to let the OSS Api Proxy know if ACM is used to authorize access to the given object. If this authorization is used by your service, then you must provide the name of the namespace you want to validate access control policies against.
     * @param {string} [xAdsAcmCheckGroups] Informs the OSS Api Proxy know if your service requires ACM authorization to also validate against Oxygen groups. If so, you must pass this header with a value of &#x60;&#x60;true&#x60;&#x60;. Otherwise, the assumption is that checking authorization against Oxygen groups is not required.
     * @param {string} [xAdsAcmGroups] Use this header to pass the Oxygen groups you want the OSS Api Proxy to use for group validation for the given user in the OAuth2 token.
     * @param {With} [_with]
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    getObjectDetails(accessToken, bucketKey, objectKey, ifModifiedSince, _with, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into getObjectDetails ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).getObjectDetails(accessToken, bucketKey, objectKey, ifModifiedSince, _with, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`getObjectDetails Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`getObjectDetails Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`getObjectDetails Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`getObjectDetails Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`getObjectDetails Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Returns a list objects in the specified bucket.   Only the application that owns the bucket can call this operation. All other applications that call this operation will receive a \"403 Forbidden\" error.
     * @summary List Objects
     * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
     * @param {number} [limit] The number of items you want per page. Acceptable values &#x3D; 1-100. Default &#x3D; 10.
     * @param {string} [beginsWith] Filters the results by the value you specify. Only the objects with their &#x60;&#x60;objectKey&#x60;&#x60; beginning with the specified string are returned.
     * @param {string} [startAt] The ID of the last item that was returned in the previous result set.  It enables the system to return subsequent items starting from the next one after the specified ID.
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    getObjects(accessToken, bucketKey, limit, beginsWith, startAt, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into getObjects ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).getObjects(accessToken, bucketKey, limit, beginsWith, startAt, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`getObjects Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`getObjects Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`getObjects Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`getObjects Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`getObjects Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Downloads an object using an OSS signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/)  contains the ``hash`` URI parameter as well.
     * @summary Download Object Using Signed URL
     * @param {string} hash The ID component of the signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains &#x60;&#x60;hash&#x60;&#x60; as a URI parameter.
     * @param {string} [range] The byte range to download, specified in the form &#x60;&#x60;bytes&#x3D;&lt;START_BYTE&gt;-&lt;END_BYTE&gt;&#x60;&#x60;.
     * @param {string} [ifNoneMatch] The last known ETag value of the object. OSS returns the requested data  only if the &#x60;&#x60;If-None-Match&#x60;&#x60; header differs from the ETag value of the object on OSS, which indicates that the object on OSS is newer. If not, it returns a 304 \&quot;Not Modified\&quot; HTTP status.
     * @param {string} [ifModifiedSince] A timestamp in the HTTP date format (Mon, DD Month YYYY HH:MM:SS GMT). The requested data is returned only if the object has been modified since the specified timestamp. If not, a 304 (Not Modified) HTTP status is returned.
     * @param {string} [acceptEncoding] The compression format your prefer to receive the data. Possible values are:  - &#x60;&#x60;gzip&#x60;&#x60; - Use the gzip format  **Note:** You cannot use &#x60;&#x60;Accept-Encoding:gzip&#x60;&#x60; with a Range header containing an end byte range. OSS will not honor the End byte range if &#x60;&#x60;Accept-Encoding: gzip&#x60;&#x60; header is used.
     * @param {Region} [region] Specifies where the bucket containing the object stored. Possible values are:  - &#x60;&#x60;US&#x60;&#x60; - (Default) Data center for the US region. - &#x60;&#x60;EMEA&#x60;&#x60; - Data center for the European Union, Middle East, and Africa. - &#x60;&#x60;APAC&#x60;&#x60; -  (Beta) Data center for Australia.  **Note:** Beta features are subject to change. Please do not use in production environments.
     * @param {string} [responseContentDisposition] The value of the Content-Disposition header you want to receive when you download the object using the signed URL. If you do not specify a value, the Content-Disposition header defaults to the value stored with OSS.
     * @param {string} [responseContentType] The value of the Content-Type header you want to receive when you download the object using the signed URL. If you do not specify a value, the Content-Type header defaults to the value stored with OSS.
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    getSignedResource(accessToken, hash, range, ifNoneMatch, ifModifiedSince, acceptEncoding, region, responseContentDisposition, responseContentType, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into getSignedResource ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).getSignedResource(accessToken, hash, range, ifNoneMatch, ifModifiedSince, acceptEncoding, region, responseContentDisposition, responseContentType, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`getSignedResource Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`getSignedResource Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`getSignedResource Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`getSignedResource Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`getSignedResource Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Gets a signed URL to download an object directly from S3, bypassing OSS servers. This signed URL expires in 2 minutes by default, but you can change this duration if needed.  You must start the download before the signed URL expires. The download itself can take longer. If the download fails after the validity period of the signed URL has elapsed, you can call this operation again to obtain a fresh signed URL.  Only applications that have read access to the object can call this operation.     You can use range headers with the signed download URL to download the object in chunks. This ability lets you download chunks in parallel, which can result in faster downloads.  If the object you want to download was uploaded in chunks and is still assembling on OSS, you will receive multiple S3 URLs instead of just one. Each URL will point to a chunk. If you prefer to receive a single URL, set the ``public-resource-fallback`` query parameter to ``true``. This setting will make OSS fallback to returning a single signed OSS URL, if assembling is still in progress.   In addition to this operation that generates S3 signed URLs, OSS provides an operation to generate OSS signed URLs. S3 signed URLs allow direct upload/download from S3 but are restricted to bucket owners. OSS signed URLs also allow upload/download and can be configured for access by other applications, making them suitable for sharing objects across applications.
     * @summary Generate Signed S3 Download URL
     * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
     * @param {string} objectKey The URL-encoded human friendly name of the object.
     * @param {string} [ifNoneMatch] The last known ETag value of the object. OSS returns the signed URL only if the &#x60;&#x60;If-None-Match&#x60;&#x60; header differs from the ETag value of the object on S3. If not, it returns a 304 \&quot;Not Modified\&quot; HTTP status.
     * @param {string} [ifModifiedSince] A timestamp in the HTTP date format (Mon, DD Month YYYY HH:MM:SS GMT). The signed URL is returned only if the object has been modified since the specified timestamp. If not, a 304 (Not Modified) HTTP status is returned.
     * @param {string} [xAdsAcmScopes] Optional OSS-compliant scope reference to increase bucket search performance
     * @param {string} [responseContentType] The value of the Content-Type header you want to receive when you download the object using the signed URL. If you do not specify a value, the Content-Type header defaults to the value stored with OSS.
     * @param {string} [responseContentDisposition] The value of the Content-Disposition header you want to receive when you download the object using the signed URL. If you do not specify a value, the Content-Disposition header defaults to the value stored with OSS.
     * @param {string} [responseCacheControl] The value of the Cache-Control header you want to receive when you download the object using the signed URL. If you do not specify a value, the Cache-Control header defaults to the value stored with OSS.
     * @param {boolean} [publicResourceFallback] Specifies how to return the signed URLs, in case the object was uploaded in chunks, and assembling of chunks is not yet complete.  - &#x60;&#x60;true&#x60;&#x60; : Return a single signed OSS URL. - &#x60;&#x60;false&#x60;&#x60; : (Default) Return multiple signed S3 URLs, where each URL points to a chunk.
     * @param {number} [minutesExpiration] The time window (in minutes) the signed URL will remain usable. Acceptable values &#x3D; 1-60 minutes. Default &#x3D; 2 minutes.  **Tip:** Use the smallest possible time window to minimize exposure of the signed URL.
     * @param {boolean} [useCdn] &#x60;&#x60;true&#x60;&#x60; : Returns a URL that points to a CloudFront edge location.  &#x60;&#x60;false&#x60;&#x60; : (Default) Returns a URL that points directly to the S3 object.
     * @param {boolean} [redirect] Generates a HTTP redirection response (Temporary Redirect 307​) using the generated URL.
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    signedS3Download(accessToken, bucketKey, objectKey, ifNoneMatch, ifModifiedSince, responseContentType, responseContentDisposition, responseCacheControl, publicResourceFallback, minutesExpiration, useCdn, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into signedS3Download ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).signedS3Download(accessToken, bucketKey, objectKey, ifNoneMatch, ifModifiedSince, responseContentType, responseContentDisposition, responseCacheControl, publicResourceFallback, minutesExpiration, useCdn, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`signedS3Download Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`signedS3Download Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`signedS3Download Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`signedS3Download Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`signedS3Download Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Gets a signed URL to upload an object directly to S3, bypassing OSS servers. You can also request an array of signed URLs which lets you upload an object in chunks.  This signed URL expires in 2 minutes by default, but you can change this duration if needed.  You must start the upload before the signed URL expires. The upload itself can take longer. If the upload fails after the validity period of the signed URL has elapsed, you can call this operation again to obtain a fresh signed URL (or an array of signed URLs as the case may be). However, you must use the same ``uploadKey`` that was returned when you originally called this operation.   Only applications that own the bucket can call this operation.  **Note:** Once you upload all chunks you must call the [Complete Upload to S3 Signed URL](/en/docs/data/v2/reference/http/buckets-:bucketKey-objects-:objectKey-signeds3upload-POST/) operation to indicate completion. This instructs OSS to assemble the chunks and reconstitute the object on OSS. You must call this operation even when using a single signed URL.   In addition to this operation that generates S3 signed URLs, OSS provides an operation to generate OSS signed URLs. S3 signed URLs allow direct upload/download from S3 but are restricted to bucket owners. OSS signed URLs also allow upload/download and can be configured for access by other applications, making them suitable for sharing objects across applications.
     * @summary Generate Signed S3 Upload URL
     * @param {string} bucketKey The bucket key of the bucket that contains the objects you are operating on.
     * @param {string} objectKey The URL-encoded human friendly name of the object.
     * @param {string} [xAdsAcmScopes] Optional OSS-compliant scope reference to increase bucket search performance
     * @param {number} [parts] The number of parts you intend to chunk the object for uploading. OSS will return that many signed URLs, one URL for each chunk. If you do not specify a value you\&#39;ll get only one URL to upload the entire object.
     * @param {number} [firstPart] The index of the first chunk to be uploaded.
     * @param {string} [uploadKey] The &#x60;&#x60;uploadKey&#x60;&#x60; of a previously-initiated upload, in order to request more chunk upload URLs for the same upload. If you do not specify a value, OSS will initiate a new upload entirely.
     * @param {number} [minutesExpiration] The time window (in minutes) the signed URL will remain usable. Acceptable values &#x3D; 1-60 minutes. Default &#x3D; 2 minutes.  **Tip:** Use the smallest possible time window to minimize exposure of the signed URL.
     * @param {boolean} [useAcceleration] &#x60;&#x60;true&#x60;&#x60; : (Default) Generates a faster S3 signed URL using Transfer Acceleration.  &#x60;&#x60;false&#x60;&#x60;: Generates a standard S3 signed URL.
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    signedS3Upload(accessToken, bucketKey, objectKey, parts, firstPart, uploadKey, minutesExpiration, useAcceleration, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into signedS3Upload ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).signedS3Upload(accessToken, bucketKey, objectKey, parts, firstPart, uploadKey, minutesExpiration, useAcceleration, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`signedS3Upload Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`signedS3Upload Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`signedS3Upload Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`signedS3Upload Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`signedS3Upload Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Replaces an object that already exists on OSS, using an OSS signed URL.   The signed URL must fulfil the following conditions:  - The signed URL is valid (it has not expired as yet). - It was generated with ``write`` or ``readwrite`` for the ``access`` parameter.
     * @summary Replace Object Using Signed URL
     * @param {string} hash The ID component of the signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains &#x60;&#x60;hash&#x60;&#x60; as a URI parameter.
     * @param {number} contentLength The size of the data contained in the request body, in bytes.
     * @param {File} body The object to upload.
     * @param {string} [contentType] The MIME type of the object to upload; can be any type except \&#39;multipart/form-data\&#39;. This can be omitted, but we recommend adding it.
     * @param {string} [contentDisposition] The suggested file name to use when this object is downloaded as a file.
     * @param {Region} [xAdsRegion] Specifies where the bucket containing the object stored. Possible values are:  - &#x60;&#x60;US&#x60;&#x60; - (Default) Data center for the US region. - &#x60;&#x60;EMEA&#x60;&#x60; - Data center for the European Union, Middle East, and Africa. - &#x60;&#x60;APAC&#x60;&#x60; -  (Beta) Data center for Australia.  **Note:** Beta features are subject to change. Please do not use in production environments.
     * @param {string} [ifMatch] The current value of the &#x60;&#x60;sha1&#x60;&#x60; attribute of the object you want to replace. OSS checks the &#x60;&#x60;If-Match&#x60;&#x60; header against the &#x60;&#x60;sha1&#x60;&#x60; attribute of the object in OSS. If they match, OSS allows the object to be overwritten. Otherwise, it means that the object on OSS has been modified since you retrieved the &#x60;&#x60;sha1&#x60;&#x60; and the request fails.
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    uploadSignedResource(accessToken, hash, contentLength, body, contentType, contentDisposition, xAdsRegion, ifMatch, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into uploadSignedResource ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).uploadSignedResource(accessToken, hash, contentLength, body, contentType, contentDisposition, xAdsRegion, ifMatch, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`uploadSignedResource Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`uploadSignedResource Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`uploadSignedResource Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`uploadSignedResource Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`uploadSignedResource Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
    /**
     * Performs a resumable upload using an OSS signed URL. Use this operation to upload an object in chunks.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains the ``hash`` as a URI parameter.
     * @summary Upload Object Using Signed URL
     * @param {string} hash The ID component of the signed URL.  **Note:** The signed URL returned by [Generate OSS Signed URL](/en/docs/data/v2/reference/http/signedresources-:id-GET/) contains &#x60;&#x60;hash&#x60;&#x60; as a URI parameter.
     * @param {string} contentRange The byte range to upload, specified in the form &#x60;&#x60;bytes&#x3D;&lt;START_BYTE&gt;-&lt;END_BYTE&gt;&#x60;&#x60;.
     * @param {string} sessionId An ID to uniquely identify the file upload session.
     * @param {File} body The chunk to upload.
     * @param {string} [contentType] The MIME type of the object to upload; can be any type except \&#39;multipart/form-data\&#39;. This can be omitted, but we recommend adding it.
     * @param {string} [contentDisposition] The suggested file name to use when this object is downloaded as a file.
     * @param {Region} [xAdsRegion] Specifies where the bucket containing the object stored. Possible values are:  - &#x60;&#x60;US&#x60;&#x60; - (Default) Data center for the US region. - &#x60;&#x60;EMEA&#x60;&#x60; - Data center for the European Union, Middle East, and Africa. - &#x60;&#x60;APAC&#x60;&#x60; -  (Beta) Data center for Australia.  **Note:** Beta features are subject to change. Please do not use in production environments.
     * @param accessToken bearer access token
     * @param {*} [options] Override http request option.
     * @throws {RequiredError}
     * @memberof ObjectsApi
     */
    uploadSignedResourcesChunk(accessToken, hash, contentRange, sessionId, body, contentType, contentDisposition, xAdsRegion, options) {
        return __awaiter(this, void 0, void 0, function* () {
            this.logger.logInfo("Entered into uploadSignedResourcesChunk ");
            try {
                const request = yield (0, exports.ObjectsApiFp)(this.sdkManager).uploadSignedResourcesChunk(accessToken, hash, contentRange, sessionId, body, contentType, contentDisposition, xAdsRegion, options);
                const response = yield request(this.axios);
                this.logger.logInfo(`uploadSignedResourcesChunk Request completed successfully with status code: ${response.status}`);
                return new autodesk_sdkmanager_1.ApiResponse(response, response.data);
            }
            catch (error) {
                if (error.response) {
                    const errorMessage = JSON.stringify(error.response.data);
                    this.logger.logError(`uploadSignedResourcesChunk Request failed with status : ${error.response.status} and statusText : ${error.response.statusText} and error message: ${errorMessage}`);
                    throw new base_1.OssApiError(`uploadSignedResourcesChunk Request failed with status : ${error.response.status} and error message: ${errorMessage}`, error);
                }
                else if (error.request) {
                    this.logger.logError(`uploadSignedResourcesChunk Request failed with no response received: ${error.request}`);
                    throw new base_1.OssApiError(`uploadSignedResourcesChunk Request failed with no response received: ${error.request}`, error);
                }
                throw error;
            }
        });
    }
}
exports.ObjectsApi = ObjectsApi;
